### PR TITLE
Extend killswitch functionality to fields within events

### PR DESCRIPTION
**Pull request by A-UNDERSCORE-D**
This PR adds the ability for killswitches to simply modify data either before or after processing to prevent data leakage.

Fields can be deleted, redacted (set to the string "REDACTED") or directly set with values given. 
If a killswitch has any kind of action, it will be attempted, and if it succeeds, the event processing with continue with modified data. Thus a simple killswitch with no additional rules simply blocks processing entirely.

### CODE REVIEW

**Comment by A-UNDERSCORE-D** [comment]
This is untested and needs unit tests, dont merge until then.

**Comment by A-UNDERSCORE-D** [comment]
Tests added and passed -- good for review, then we need to add the new files before merging

**Review by Athanasius — CHANGES_REQUESTED** [review]
I've only looked at the docs so far, but might as well get you moving on these changes whilst I look at the actual code.

File: docs/Killswitches.md
Comment (A-UNDERSCORE-D) [reply docs/Killswitches.md:41]: include none of them, `{}`

File: docs/Killswitches.md
Comment (A-UNDERSCORE-D) [reply docs/Killswitches.md:66]: yeah it just upgrades the format internally, it wont change behaviour

File: docs/Killswitches.md
Comment (A-UNDERSCORE-D) [reply docs/Killswitches.md:77]: > Also, you're using | just to show here that it could apply to any of the core plugins, rather than as a possible syntax in the actual killswitch right?
Yeah, I have a bad habit of sticking regex where it doesnt belong

>PyCharm's MK rendering barfs on this, still interpreting the | as a table column separator.
sounds like a bug report to me :D but yes Ill fix it

File: killswitch.py
Comment (Athanasius) [inline killswitch.py:213]: flake8: no blank line after function docstring (yes, I hate the inconsistency with *requiring* such after class docstrings).

File: killswitch.py
Comment (Athanasius) [inline killswitch.py:85]: This doesn't catch where a *negative* index is too (negatively) large.

File: killswitch.py
Comment (Athanasius) [inline killswitch.py:92]: Negative index also an issue here.

We could decide to just not support them, but I could see the usefulness given it's valid python slicing syntax.

File: killswitch.py
Comment (Athanasius) [inline killswitch.py:102]: docstrings & comments should stay to 80 columns at most

File: killswitch.py
Comment (Athanasius) [inline killswitch.py:223]: I had to look up the semantic_version docs *and* run a test to be sure, so add a comment here that, yes, this 'in' syntax does support things like '<1.2.3' for range comparisons.

Comment (Athanasius) [reply killswitch.py:223]: NB: This could be in Killswitches.md#versions rather than in the code.  All it does is say it's a SimpleSpec, and examples are simple specific versions, not comparison versions.  We need to document that range comparison versions are supported.

File: killswitch.py
Comment (Athanasius) [inline killswitch.py:310]: I'm antsy about this remote-data request not being in a thread, but on reflection:

1. If we have core-code killswitches then we **really** *do* want to load them before running any code they might affect.
2. For now I doubt any plugin is using them, and at very worst we'll take a 10 second hit for core code plus each plugin attempting to load killswitches.

Resolving this to ensure core code always runs with it kill switches is mostly moot because what are we going to do if they're not loaded yet?  Hang around for them to be loaded!

For plugins we could document how plugins can load them in a thread (provide a helper?) and also how to ensure they don't process data until the load/check is performed.

**Comment by Athanasius** [comment]
We could use a script to test a killswitch set for what it will actually do.  At the very least ensure it parses OK without running the full program (so can be tested without GUI-ness available).

**Comment by Athanasius** [comment]
Killswitch.py:380:
```python
    logger.info(f'Kill switches last updated {last_updated}')
```
is fine if it was only core killswitches, but this will be ambiguous for extra/plugin killswitches having been loaded.

**Comment by Athanasius** [comment]
Killswitch.py:425
```
def setup_main_list():
    """
    Set up the global set of kill switches for querying.

    Plugins should NOT call this EVER.
```
I don't suppose we want to play silly buggers with checking the call stack to see if a plugin has done that ....  Probably not worth it.

**Review by Athanasius — CHANGES_REQUESTED** [review]
OK, first full review done.  As you can tell from the out-of-review comments this might be the first time I took a serious look at the code at all :P

**Comment by Athanasius** [comment]
> We could use a script to test a killswitch set for what it will actually do. At the very least ensure it parses OK without running the full program (so can be tested without GUI-ness available).

I guess maybe tests/killswitch.py/test_apply could be altered to allow for an input file ?

**Comment by Athanasius** [comment]
> > We could use a script to test a killswitch set for what it will actually do. At the very least ensure it parses OK without running the full program (so can be tested without GUI-ness available).
> 
> I guess maybe tests/killswitch.py/test_apply could be altered to allow for an input file ?

Failing that *some* script, because then we could potentially do something with a push action that only works on the `releases` branch to check the current killswitches file is actually parsable.  That and I could probably lever it into the pre-commit config as well.

**Comment by A-UNDERSCORE-D** [comment]
> Killswitch.py:425
> 
> ```
> def setup_main_list():
>     """
>     Set up the global set of kill switches for querying.
> 
>     Plugins should NOT call this EVER.
> ```
> 
> I don't suppose we want to play silly buggers with checking the call stack to see if a plugin has done that .... Probably not worth it.

Not at all.

>We could use a script to test a killswitch set for what it will actually do. At the very least ensure it parses OK without running the full program (so can be tested without GUI-ness available).

What exactly would you like, just verifying that everything parses correctly?

**Comment by Athanasius** [comment]
> > We could use a script to test a killswitch set for what it will actually do. At the very least ensure it parses OK without running the full program (so can be tested without GUI-ness available).
> 
> What exactly would you like, just verifying that everything parses correctly?

1. The bare minimum is at least that, for the purposes of verifying the file before making it live (rather than a type/thinko meaning it's invalid and we in effect lose all the prior killswitches as a result).
2. Beyond that something to ensure that, e.g. an id is at least citing a plugin that *exists* to catch typos.
3. Beyond that something that literally tells you "disables plugins.edsm.worker.scan", "forces plugins.edsm.journal.event.horizons to True" etc.

**Comment by A-UNDERSCORE-D** [comment]
While working on the requested script I realised that two of our four plugins havent been updated to support the new stuff, thats additionally blocking this

File: killswitch.py
Comment (A-UNDERSCORE-D) [reply killswitch.py:52]: Order is now explicitly defined, and tested

File: killswitch.py
Comment (A-UNDERSCORE-D) [reply killswitch.py:222]: Clarified in docs

File: killswitch.py
Comment (A-UNDERSCORE-D) [reply killswitch.py:310]: Added a oneliner helper and doccumented it. As for core code, yeah, I want us to sit and wait, plugins can just defer anything they send until its there (or do it on-thread and block)

**Comment by A-UNDERSCORE-D** [comment]
Killswitch test has been added, plugins updated.

File: docs/Releasing.md
Comment (Athanasius) [inline docs/Releasing.md:244]: Mention the testing script here, although we probably want it adding into the pre-commit config as well.

File: killswitch.py
Comment (Athanasius) [inline killswitch.py:257]: I would word this as "A two value tuple consisting of ...".

File: plugins/edsm.py
Comment (Athanasius) [inline plugins/edsm.py:688]: This can use trace_if(...) now.

File: plugins/edsm.py
Comment (Athanasius) [inline plugins/edsm.py:695]: trace_if

File: docs/Killswitches.md
Comment (Athanasius) [inline docs/Killswitches.md:86]: This `occur--Any` just looks weird.  I'd use:

> No behaviour changes will occur, any killswitches...

If not:

> No behaviour changes will occur.  Any killswitches...

or were you going for:

> No behaviour changes will occur -- any killswitches...

?

File: killswitch.py
Comment (Athanasius) [inline killswitch.py]: declare return type ?

File: docs/Killswitches.md
Comment (Athanasius) [inline docs/Killswitches.md:104]: Line over 80 chars long.  Actually now I check in VIM to be sure, the whole file could do with a pass to bring all non-table/link stuff within that limit.

**Review by Athanasius — CHANGES_REQUESTED** [review]
Just a few niggles in docs and formatting therein.

File: docs/Killswitches.md
Comment (A-UNDERSCORE-D) [reply docs/Killswitches.md:86]: Sorry bad habit, I use -- as an em-dash which shouldnt have spaces, but a period works fine there as well

File: killswitch.py
Comment (A-UNDERSCORE-D) [reply killswitch.py]: Done where possible

File: plugins/edsm.py
Comment (A-UNDERSCORE-D) [reply plugins/edsm.py]: Resolved with rebase

File: plugins/edsm.py
Comment (A-UNDERSCORE-D) [reply plugins/edsm.py]: Resolved with rebase

File: killswitch.py
Comment (A-UNDERSCORE-D) [reply killswitch.py]: I *think* I hadnt pushed changes or something for this. They should be there now.

File: docs/Killswitches.md
Comment (A-UNDERSCORE-D) [reply docs/Killswitches.md]: Resolved in all the files I touched docs wise

File: killswitch.py
Comment (Athanasius) [reply killswitch.py:93]: 1. Either this needs to allow for setting a non-existent index that isn't an append.

2. OR Killswitches.md needs to document that an index on a set must either be an existing one, or only a simple append.

3. OR this function needs to be marked as explicitly possibly raising `IndexError`.

2 or 3 needs a test put in to ensure that happens as documented.

I added:

```python
        (['test append'], '2', '', 'yay', ['test append', 'yay']),
```
to the 'test_apply()' parameters in `test_apply.py` and it blew up with an IndexError.

File: tests/killswitch.py/test_apply.py
Comment (Athanasius) [inline tests/killswitch.py/test_apply.py:84]: This is essentially the same test as `multilevel.keys` one two lines above.

Comment (A-UNDERSCORE-D) [reply tests/killswitch.py/test_apply.py:84]: I was apparently half asleep while writing those. I dont think theres any harm in leaving it?

File: docs/Killswitches.md
Comment (Athanasius) [inline docs/Killswitches.md:74]: As I went searching yesterday, add a note that:

```
To append you will need to use `len(array/set/whatever)`, there's no way to do it just with indexing.
```

I wish `-0` was special in that way, but obviously all of python's parsing would need the concept of 'negative zero' for that to work.

Comment (Athanasius) [reply docs/Killswitches.md:74]: Ah never mind, you covered this below in the Sets section.

File: docs/Killswitches.md
Comment (Athanasius) [inline docs/Killswitches.md:117]: list/array and dict ?

File: killswitch.py
Comment (Athanasius) [inline killswitch.py:71]:
```
NB: Other exceptions may be raised, i.e. IndexError if supplying a bad index.
```

My point here is to not have someone who's not read all the code/docs think that this can *only* raise ValueError.

File: tests/killswitch.py/test_apply.py
Comment (Athanasius) [reply tests/killswitch.py/test_apply.py:84]: No harm, other than someone checking this wondering *why* there's essentially the same test twice, and going down a rabbithole trying to figure out how it's different.  So remove this second one please.

**Pull request merged**
